File: bloom_filter_test.jl
# bloom_filter_test.jl
#
# curl -fsSL https://install.julialang.org | sh
#
# julia bloom_filter_test.jl

using SHA  # Standard library for cryptographic hashing

struct BloomFilter
    bit_array::BitVector
    num_hashes::Int
    size::Int
end

# Constructor to initialize our filter with zeros
function BloomFilter(size::Int, num_hashes::Int)
    return BloomFilter(falses(size), num_hashes, size)
end

"""
    get_hashes(item::String, k::Int, max_val::Int)

Generates `k` distinct bit-array indices for a given string using 
salting and SHA-256 to simulate multiple independent hash functions.
"""
function get_hashes(item::String, k::Int, max_val::Int)
    indices = Int[]
    for i in 1:k
        # Salt the string with the current index loop to get diverse hashes
        salted_bytes = sha256(item * string(i))
        
        # Take the first 8 bytes of the hash and convert to an integer
        raw_int = reinterpret(UInt64, salted_bytes[1:8])[1]
        
        # Map it to our bit array boundaries (1-indexed for Julia arrays)
        push!(indices, (raw_int % max_val) + 1)
    end
    return indices
end

"""
    insert!(bf::BloomFilter, item::String)

Flips the calculated hash bits to true.
"""
function insert!(bf::BloomFilter, item::String)
    for idx in get_hashes(item, bf.num_hashes, bf.size)
        bf.bit_array[idx] = true
    end
end

"""
    check(bf::BloomFilter, item::String)::Bool

Returns true if the item MIGHT be in the set, false if it DEFINITELY is not.
"""
function check(bf::BloomFilter, item::String)::Bool
    for idx in get_hashes(item, bf.num_hashes, bf.size)
        if !bf.bit_array[idx]
            return false # Found a 0, so it absolutely never entered the set
        end
    end
    return true # All bits were 1, might be a false positive!
end

# Run the test simulation
function main()
    # Create a small 1000-bit array utilizing 4 hash passes
    bf = BloomFilter(1000, 4)
    
    # 1. Populate the filter with Old Testament names
    println("--- Inserting Old Testament Records into Bloom Filter ---")
    old_testament_records = ["abraham_patriarch", "moses_prophet", "david_king", "esther_queen"]
    for person in old_testament_records
        insert!(bf, person)
        println("Registered: ", person)
    end
    
    println("\n--- Querying the Registry ---")
    
    # Test an item that definitely exists in the filter
    test_present = "moses_prophet"
    println("Checking '$(test_present)' (Should be true): ", check(bf, test_present))
    
    # Test an item that belongs to a completely different era/text (New Testament)
    test_absent = "paul_apostle"
    println("Checking '$(test_absent)' (Should be false): ", check(bf, test_absent))
    
    # Let's inspect the underlying BitVector optimization
    active_bits = sum(bf.bit_array)
    println("\nFilter Density: ", active_bits, " out of ", bf.size, " bits flipped to true.")
end

main()

File: bloom_filter_test.lua
-- brew install lua
--
-- lua -v
--
-- lua bloom_filter_test.lua

local BloomFilter = {}
BloomFilter.__index = BloomFilter

-- Pure Lua FNV-1a 32-bit hash implementation
local function fnv1a_32(data, seed)
    local prime = 16777619
    local hash = seed or 2166136261
    
    for i = 1, #data do
        local byte = string.byte(data, i)
        -- XOR the lowest 8 bits of the hash with the current byte
        hash = (hash ~ byte) & 0xFFFFFFFF
        -- Multiply by FNV prime and maintain 32-bit integer limits
        hash = (hash * prime) & 0xFFFFFFFF
    end
    return hash
end

-- Constructor
-- @param size: The total number of bits in the bit array (m)
-- @param num_hashes: The number of hash functions to use (k)
function BloomFilter.new(size, num_hashes)
    local self = setmetatable({}, BloomFilter)
    self.size = size
    self.num_hashes = num_hashes
    
    -- Initialize the bit array. 
    -- In Lua, tables act as sparse arrays, so initializing to false is highly memory efficient.
    self.bit_array = {}
    for i = 0, size - 1 do
        self.bit_array[i] = false
    end
    
    return self
end

-- Add an item to the Bloom Filter
function BloomFilter:add(item)
    item = tostring(item)
    for i = 1, self.num_hashes do
        -- Use the loop index as a unique seed to simulate independent hash functions
        local hash = fnv1a_32(item, i * 1103515245)
        local index = hash % self.size
        self.bit_array[index] = true
    end
end

-- Check if an item is likely in the Bloom Filter
-- @return boolean: false means "Definitely Not", true means "Probably"
function BloomFilter:check(item)
    item = tostring(item)
    for i = 1, self.num_hashes do
        local hash = fnv1a_32(item, i * 1103515245)
        local index = hash % self.size
        
        -- If any single bit is false, the item has definitively never been added
        if not self.bit_array[index] then
            return false
        end
    end
    return true
end

--- ========================================================================
--- DEMO EXECUTION
--- ========================================================================

print("--- Initializing Bloom Filter (Size: 1000 bits, Hashes: 5) ---")
-- Allocating a small bit-array room for demo visibility
local bf = BloomFilter.new(1000, 5)

-- 1. Populate the filter with a known allowed-list (e.g., trusted IP addresses or safe URLs)
local trusted_ips = {
    "192.168.1.1",
    "10.0.0.50",
    "172.16.254.1",
    "8.8.8.8"
}

print("\n[+] Adding trusted IPs to the filter...")
for _, ip in ipairs(trusted_ips) do
    bf:add(ip)
    print("    Added: " .. ip)
end

-- 2. Test Membership Verification
print("\n--- Testing Membership (Definitive Negatives & Probable Positives) ---")

local test_cases = {
    "192.168.1.1",   -- Should be True (Was added)
    "8.8.8.8",       -- Should be True (Was added)
    "192.168.1.99",  -- Should be False (Never added)
    "10.0.0.1",      -- Should be False (Never added)
    "200.100.50.25"  -- Should be False (Never added)
}

for _, ip in ipairs(test_cases) do
    local result = bf:check(ip)
    if result then
        print(string.format("--> Is '%s' trusted? PROBABLY YES", ip))
    else
        print(string.format("--> Is '%s' trusted? DEFINITELY NO", ip))
    end
end